File: examples/spring-boot-minimal/src/main/java/example/springframework/boot/minimal/ValidationExceptionHandler.java
package example.springframework.boot.minimal;

import static java.util.Objects.requireNonNull;

import java.time.Instant;

import javax.validation.ValidationException;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import com.linecorp.armeria.common.HttpRequest;
import com.linecorp.armeria.common.HttpResponse;
import com.linecorp.armeria.common.HttpStatus;
import com.linecorp.armeria.server.ServiceRequestContext;
import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction;

/**
 * A sample exception handler which handles a {@link ValidationException}.
 */
public class ValidationExceptionHandler implements ExceptionHandlerFunction {

    @Override
    public HttpResponse handleException(ServiceRequestContext ctx, HttpRequest req, Throwable cause) {
        if (cause instanceof ValidationException) {
            final HttpStatus status = HttpStatus.BAD_REQUEST;
            return HttpResponse.ofJson(status, new ErrorResponse(status.reasonPhrase(),
                                                                 cause.getMessage(),
                                                                 req.path(),
                                                                 status.code(),
                                                                 Instant.now().toString()));
        }
        return ExceptionHandlerFunction.fallthrough();
    }

    /**
     * A sample HTTP response which is sent to a client when a {@link ValidationException} is raised.
     */
    public static class ErrorResponse {
        private final String error;
        private final String message;
        private final String path;
        private final int status;
        private final String timestamp;

        @JsonCreator
        ErrorResponse(@JsonProperty("error") String error,
                      @JsonProperty("message") String message,
                      @JsonProperty("path") String path,
                      @JsonProperty("status") int status,
                      @JsonProperty("timestamp") String timestamp) {
            this.error = requireNonNull(error, "error");
            this.message = requireNonNull(message, "message");
            this.path = requireNonNull(path, "path");
            this.status = status;
            this.timestamp = requireNonNull(timestamp, "timestamp");
        }

        @JsonProperty
        public String error() {
            return error;
        }

        @JsonProperty
        public String message() {
            return message;
        }

        @JsonProperty
        public String path() {
            return path;
        }

        @JsonProperty
        public int status() {
            return status;
        }

        @JsonProperty
        public String timestamp() {
            return timestamp;
        }
    }
}

File: examples/spring-boot-kotlin/src/main/java/example/springframework/boot/minimal/ValidationExceptionHandler.kt
package example.springframework.boot.minimal

import java.util.Objects.requireNonNull

import java.time.Instant

import javax.validation.ValidationException

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper

import com.linecorp.armeria.common.HttpRequest
import com.linecorp.armeria.common.HttpResponse
import com.linecorp.armeria.common.HttpStatus
import com.linecorp.armeria.common.MediaType
import com.linecorp.armeria.server.ServiceRequestContext
import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction

/**
 * A sample exception handler which handles a [ValidationException].
 */
class ValidationExceptionHandler : ExceptionHandlerFunction {

    private val mapper = ObjectMapper()

    override fun handleException(ctx: ServiceRequestContext, req: HttpRequest, cause: Throwable): HttpResponse {
        if (cause is ValidationException) {
            try {
                val status = HttpStatus.BAD_REQUEST
                return HttpResponse.of(status, MediaType.JSON_UTF_8, mapper.writeValueAsBytes(
                        ErrorResponse(status.reasonPhrase(),
                                cause.message ?: "Invalid Parameter",
                                req.path(),
                                status.code(),
                                Instant.now().toString())))
            } catch (e: JsonProcessingException) {
                return HttpResponse.of(HttpStatus.INTERNAL_SERVER_ERROR, MediaType.PLAIN_TEXT_UTF_8,
                        cause.message)
            }

        }
        return ExceptionHandlerFunction.fallthrough()
    }

    /**
     * A sample HTTP response which is sent to a client when a [ValidationException] is raised.
     */
    class ErrorResponse @JsonCreator
    internal constructor(@JsonProperty("error") error: String,
                         @JsonProperty("message") message: String,
                         @JsonProperty("path") path: String,
                         @param:JsonProperty("status") private val status: Int,
                         @JsonProperty("timestamp") timestamp: String) {
        private val error: String
        private val message: String
        private val path: String
        private val timestamp: String

        init {
            this.error = requireNonNull(error, "error")
            this.message = requireNonNull(message, "message")
            this.path = requireNonNull(path, "path")
            this.timestamp = requireNonNull(timestamp, "timestamp")
        }

        @JsonProperty
        fun error(): String {
            return error
        }

        @JsonProperty
        fun message(): String {
            return message
        }

        @JsonProperty
        fun path(): String {
            return path
        }

        @JsonProperty
        fun status(): Int {
            return status
        }

        @JsonProperty
        fun timestamp(): String {
            return timestamp
        }
    }
}
